### PR TITLE
k8s_auth: new k8s module for handling auth

### PR DESCRIPTION
**Pull request by mmazur**
##### SUMMARY
Some kubernetes distributions (like OpenShift) may require explicit login/logout steps before any of the `k8s*` modules can be used. This is the module that will handle those.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

modules/clustering/k8s/k8s_auth

### CODE REVIEW

**Comment by mmazur** [comment]
No code yet, just the docs/options/return values/examples to hopefully get some early input.

One thing that bugs me in the example is that with this module one needs to keep repeating all the ssl settings and `host` urls in each `k8s*` module invocation. I think the proper way to address this would be to also modify the base k8s code so that all modules accept the `k8s_auth` facts dict and take all their configs from it. Something like this (as opposed to what's in the EXAMPLES section):

```
- k8s_facts:
    k8s_auth: {{ k8s_auth }}
    kind: Pod
  register: pod_list
```

Oh, and btw, I've kept all the option names. Though I'd really like to see `ssl_ca_cert` renamed to something like `ca_path` everywhere.

Ping @fabianvf

**Comment by fabianvf** [comment]
Adding a new dict parameter that allows you to specify all your authentication options makes sense to me, @willthames any thoughts about that? Presumably we'd maintain the old parameters and just merge them into this dict in the code.

About `ssl_ca_cert`, do you know if it only works if you specify a path or can you provide one as a string? For some reason I thought it was either but I could be wrong.

**Comment by willthames** [comment]
@fabianvf - I'd prefer to avoid adding special dicts that replace common settings - we seem to have enough issues getting auth params right without introducing further complexity (unless such a change came with a rock solid test suite)

Usually when passing common settings I suggest using YAML anchors: http://willthames.github.io/2017/07/17/so-you-want-to-test-aws-modules-for-ansible.html#tasksmainyml

So here you'd do something like:

```
- k8s_auth:
     ...
   register: k8s_auth_result

- set_fact:
     &k8s_auth: "{{ k8s_auth_result }}"

- k8s_facts:
    <<: *k8s_auth
```

The downside of YAML anchors is they only work per file (but you could reuse `k8s_auth_result` in later files with the same `set_fact`)

In summary, I can see the argument for an `auth` parameter, and if an implementation came with a rigorous test suite, I'd support it, but YAML might suffice in the interim. ~And I think there's a mechanism used for GCE and AWS for reusing auth that might be better still, I'm just trying to find out what it's called, and whether we could use the same concept for k8s.~ 

With 2.8, we'll be able to use module_defaults with a group (we'd need to call the group something that doesn't conflict with an existing module, so k8s and kubernetes are out) - see https://docs.ansible.com/ansible/devel/user_guide/playbooks_module_defaults.html for the concept and https://github.com/ansible/ansible/blob/devel/lib/ansible/config/module_defaults.yml for how groups are defined - we'd need to add the k8s modules there

**Comment by mmazur** [comment]
A combination of `module_defaults` for groups and `set_facts` does indeed look like the proper way to handle this. I will use a version of `set_facts` in the example for this module to make sure people end up using it.

@willthames are you 100% certain groups can't be named the same as an existing module? The implementation of groups for `module_defaults` expects one to use `group/` prefix, so I'm not entirely sure why it's a problem for there to exist both `group/k8s` and `k8s` as different arguments to `module_defaults`.

**Comment by willthames** [comment]
@mmazur, it's not particularly well documented, as far as I can see, so I was making an assumption - using `group/k8s` would totally work for me. 

The nice thing is we don't have many modules so setting module_defaults per module isn't too much of a stretch until 2.8 is released.

**Comment by fabianvf** [comment]
For anyone else curious about defaults for module groups: https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/user_guide/module_defaults_config.rst

I really like the module defaults. We should definitely include an example using it.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (fabianvf) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py]: Should we `no_log` this as well? I still need to verify it can't be the actual cert content (rather than a path).

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (fabianvf) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py]: Do we need to guard this import?

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (fabianvf) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py]: Are f-strings available in Python2?

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (fabianvf) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py]: Maybe set this dict in the above if/else block

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (fabianvf) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py:18]: It may make more sense to call it like `k8s_oauth_token` or something, I would assume I could use this for generic authorization otherwise. Alternatively we could just make it succeed with a no-op in the case that the cluster doesn't support oauth.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (fabianvf) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py]: requests may not respect the standard CA trust store, so I think this might not be true

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (fabianvf) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py:58]: I need to verify it has to be a path and that you can't specify the cert directly.

**Review by fabianvf — COMMENTED** [review]
Looking good so far, have a couple questions/suggestions but I think (especially with module/group defaults) this will be a really nice feature.

**Comment by willthames** [comment]
Presumably that page has orphan on it because there are no incoming links. It would be good to get the module defaults documentation updated to include the groups stuff in the devel tree

**Comment by willthames** [comment]
I've raised a PR to improve the module group defaults documentation (thanks for the pointer to that page @fabianvf )

https://github.com/ansible/ansible/pull/51045

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py]: k8s/common.py says:

```
    'ssl_ca_cert': {
        'type': 'path',
    },
```
so it's definitively a path.

Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py]: Hmm, as I look at the code, I'm not sure that's even implemented…

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py]: This was only a thing I'm using for debugging to get the whole auth procedure through mitmproxy. I'll remove it anyways.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py]: It does:

```
>>> import requests
>>> requests.get('https://github.com')
<Response [200]>
```

Comment (fabianvf) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py]: I think requests uses the certs from the `certifi` package, not the system store. So if you have a custom certificate requests will not respect it.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (fabianvf) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py]: It passes it through to the Kubernetes client's Configuration object which should handle it, but I don't think this path has been tested thoroughly. I'll see if I can find some time to investigate it soon

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:18]: My thinking in naming it `k8s_auth` was that it should support all the auth mechanisms that will be used for k8s clusters now or in the future. With openshift that's only oauth, but if something somewhere started using zauth or ëauth, that support should be added to this module ideally with some autodetection (which should be trivial).

As to succeeding even when there's no oauth – that'd only be an option if the module first checked for oauth and if that was gone, tried to perform an operation with http basic auth to verify that that in fact works aka that the auth credentials are correct. (Same logic with using client cert instead of user+pwd, should someone add that eventually.) And depending on whether it got oauth or not, the return structure would either contain a token and an empty `password` field or an empty `api_key` field and valid user+pwd. So the fact/anchor creation code from the example would look like this:

```
- set_fact:
    k8s_auth_params: &k8s_auth_params
        host: "{{ k8s_auth_results.k8s_auth.host }}"
        ssl_ca_cert: "{{ k8s_auth_results.k8s_auth.ssl_ca_cert }}"
        verify_ssl: "{{ k8s_auth_results.k8s_auth.verify_ssl }}"
        api_key: "{{ k8s_auth_results.k8s_auth.api_key }}"
        username: "{{ k8s_auth_results.k8s_auth.username }}"
        password: "{{ k8s_auth_results.k8s_auth.password }}"
```

and the `k8s_auth_params` anchor could be used universally as it'd then work both with openshift and k8s clusters. Though again – that would depend on  `api_key` being non–null with openshift and `password` being non–null with http basic auth. I'm not exactly sure that doing the latter (returning a password and storing it) is a valid thing to do in ansible and whether the whole thing isn't too contrived.

@willthames, could you maybe comment as well?

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py]: Ok. And btw, CA cert chains are just public keys, so there's no need to `no_log` them.

Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py]: No, you wanted `no_log` to make sure there isn't hundreds of lines of pem files in the output. Makes sense.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py]: Yup, done.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py]: Good point, done.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py]: Damn, this isn't the first time I've made this assumption. Thanks, I've removed the doc string.

**Comment by mmazur** [comment]
Ok, I think I'm done with the coding. Everything I wanted to have working is working now. Removing the [WIP] tag.

Two things remaining:

- figuring out the name of this module
- if anybody has an idea what to do with the remaining CI errors, I'm all ears.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (willthames) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py:139]: Needs a `default: true` to match the argument spec.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (willthames) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py]: comment should be `ssl_ca_cert  # path`

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (willthames) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py]: two spaces before comment, one space after `#` - `verify_ssl  # bool`

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (willthames) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py]: remove this blank line

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (willthames) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py]: remove the space at the end of this line

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (willthames) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py]: should be `'req' + k`

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (willthames) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py:154]: use ansible.module_utils.url rather than requests

**Review by willthames — CHANGES_REQUESTED** [review]
I've added notes for all the bits that are currently breaking CI. Let me know if any remain unclear.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:139]: I've changed it, but it looks really weird to me…

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:154]: Are there any docs for it?

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:139]: Ok, please disregard that comment, fixed now.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:154]: Nope, and it doesn't support `ca_path`.

**Comment by mmazur** [comment]
Ok, I've checked `ansible.module_utils.urls.open_url()`, but it doesn't support providing one's own `ca_path`, so there's really no way for me to do a rewrite to remove `requests` dependency.

So, this is my final version (unless you want to continue the renaming discussion first) and please ignore that one CI error. ping @fabianvf @willthames

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (machacekondra) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py:92]: This brings unnecessary complexity to user, why not make it simpler, and set such fact inside the module?

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (machacekondra) [inline lib/ansible/modules/clustering/k8s/k8s_auth.py:261]: Actually, I think it would be better if such functionality is implemented inside openshift SDK, because more client codes could benefit from it, rather then other re-implement it their own way.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (willthames) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:92]: Are you suggesting using `ansible_facts` for that (which seems like a bit of a hack?). I'm not aware of another way of transferring state from one module to other modules.

Even if using `ansible_facts` (and I don't think we should) we still need to document for people with `INJECT_FACTS_AS_VARS` set to False

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:261]: Well, the call by the openshift client maintainer (Fabian) was to do this directly in the module, at least initially. If we ever get a need to move that code into the lib, that can be a separate conversation.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:92]: I'm pretty sure Ondra's suggesting returning to my initial code with the module accepting an `k8s_auth` dict as a parameter (as well as returning one) and having all of the k8s modules accept that.

Summary of the discussion around this is – the proper way to do this kind of thing going forward will be with with [module defaults for groups](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/user_guide/playbooks_module_defaults.rst#module-defaults-groups), which became available since 2.7. So the yaml anchor usage in the EXAMPLE is basically just a stop–gap for how this will end up being used properly in the future.

So while it made sense to implement a dedicated auth dict parameter for groups of modules pre–2.7, it really would not be the proper implementation nowadays.

Comment (machacekondra) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:92]: Yes, I meant using `ansible_facts`, but now I realized you can't create anchor from that I guess.
Yeah the module defaults solve this quite nice, actually, so forget my comment.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (fabianvf) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:261]: Yeah, I think it's best to do it here for now, but I could see incorporating this back into the client later.

File: lib/ansible/modules/clustering/k8s/k8s_auth.py
Comment (fabianvf) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:18]: makes sense to me, I think in the long term it would be cool to be able to dump in all your parameters and get back a valid param block that you can add to module_defaults.

Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:18]: Ok, I've added a comment about this with a link to this particular part of the discussion, so that future changes stay in line with this general design.

As to implementing http basic auth – I've only gotten the go ahead to implement openshift for now (and the description for the module reflects that). If I get a green light from my team to take care of doing the group module defaults thing, I'll make sure to also add http basic auth fallback to this module and have this be as universal and easy to use as possible.

Comment (willthames) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:18]: By group module_defaults do you mean https://github.com/ansible/ansible/pull/51093 ?

Comment (mmazur) [reply lib/ansible/modules/clustering/k8s/k8s_auth.py:18]: Ah, great, having this exist will make my job much easier.

**Comment by mmazur** [comment]
Ok, pushed a version that ignores the "don't import `requests`" error, since that's completely not an option right now. Now need someone with repo superpowers to plz merge.

**Comment by mattclay** [comment]
@mmazur Docs fragments have moved, so this needs to be rebased.

**Comment by mmazur** [comment]
@mattclay done.

**Comment by fabianvf** [comment]
CI failures seem unrelated, can someone rekick this?

**Comment by mattclay** [comment]
Failed tests have been restarted.

**Comment by mmazur** [comment]
Still failing. Stuff like `"assertion": "current_time.stat.atime > previous_time.stat.atime"` suggests to me that the CI vms are wonky.

**Review by willthames — APPROVED** [review]

**Comment by mmazur** [comment]
@mattclay can you run the tests again, maybe CI got stable(r) in the mean time.

**Comment by fabianvf** [comment]
shipit

**Comment by fabianvf** [comment]
bot_status

**Comment by maxamillion** [comment]
rebuild_merge

**Comment by mmazur** [comment]
ping

**Comment by maxamillion** [comment]
rebuild_merge

**Comment by mmazur** [comment]
I'm very confused on what's happening. Maybe this will work:

sudo_merge_force

**Comment by fabianvf** [comment]
bot_status

**Comment by fabianvf** [comment]
shipit

**Comment by maxamillion** [comment]
!needs_revision

**Comment by mmazur** [comment]
@mattclay k8s maintainers' shipits aren't enough here. Can you push this?

**Pull request merged**
